In [78]:
#from IPython.display import display, HTML
#display(HTML("<style>.container { width:100% !important; }</style>"))
# %matplotlib notebook
%matplotlib ipympl 
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import axes3d
import matplotlib.animation as animation
from tqdm import tqdm

In [79]:
def f2(x,y):
    return x**4 + y**4 - 2*(x-y)**2
    #return 14*x**2 - 2*x**3 + 2*y**2 + 4*x*y
Dx = lambda f,x,y,h=1e-5: (f(x+h,y) - f(x-h,y))/(2*h)
Dy = lambda f,x,y,h=1e-5: (f(x,y+h) - f(x,y-h))/(2*h)
x0, y0 = 0.5,0.1
Gradient = lambda f,x,y: np.array([Dx(f,x,y),Dy(f,x,y)])
Gradient(f2,x0,y0)


  

array([-1.1  ,  1.604])

In [80]:
#Algortimo Original
def Minimizer(f, N = 100, gamma = 0.01):
    
    r = np.zeros((N,2))
    r[0] = np.random.uniform(-5.,5.,size=2)
    
    Grad = np.zeros((N,2))
    Grad[0] = Gradient(f,r[0,0],r[0,1])
    
    # We save the gradient in each step

    for i in tqdm(range(1,N)):
        r[i] = r[i-1] - gamma*Gradient(f,r[i-1,0],r[i-1,1])
        Grad[i] = Gradient(f,r[i-1,0],r[i-1,1])
        
        
    return r,Grad

In [81]:
#Algoritmo con momentun 

def Minimizer_Momentum(f, N=100, alpha=0.001, beta=0.6):
    r = np.zeros((N, 2))
    r[0] = np.array([0, 3])
    
    Grad = np.zeros((N, 2))
    Grad[0] = Gradient(f, r[0, 0], r[0, 1])
    
    for i in tqdm(range(1, N)):
        gradient = Gradient(f, r[i - 1, 0], r[i - 1, 1])
        r[i] = r[i - 1] - alpha * gradient + beta * (r[i - 1] - r[i - 2])
        Grad[i] = gradient
    
    return r, Grad




In [82]:
#Realizamos la comparacion del numero de interaciones del algoritmo original y con momentum

# iteraciones original
N1 = 200
r_original, _ = Minimizer(f2, N1)
iterations_original = np.where(np.linalg.norm(r_original[1:] - r_original[:-1], axis=1) < 1e-6)[0][0] + 1
print("Número de iteraciones sin momentum:", iterations_original)


100%|██████████| 199/199 [00:00<00:00, 38744.21it/s]

Número de iteraciones sin momentum: 109


In [83]:
# iteraciones momentum
N2 = 1000
betas = [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]  # Establecemos unas serie de valores de coeficiente de momentum
for beta in betas:
    r_momentum, _ = Minimizer_Momentum(f2, N2, beta=beta)
    
    if len(r_momentum) >= 2:
        convergence_indices = np.where(np.linalg.norm(r_momentum[1:] - r_momentum[:-1], axis=1) < 1e-6)[0]
        if len(convergence_indices) > 0:
            iterations_momentum = convergence_indices[0] + 1
            print(f"Número de iteraciones con momentum (beta={beta}): {iterations_momentum}")
        else:
            print(f"No se alcanzó el criterio de convergencia para beta={beta}.")
    else:
        print(f"No se realizaron suficientes iteraciones para beta={beta}.")



100%|██████████| 999/999 [00:00<00:00, 59891.22it/s]


Número de iteraciones con momentum (beta=0.1): 620


100%|██████████| 999/999 [00:00<00:00, 78545.90it/s]


Número de iteraciones con momentum (beta=0.2): 552


100%|██████████| 999/999 [00:00<00:00, 65750.93it/s]


Número de iteraciones con momentum (beta=0.3): 484


100%|██████████| 999/999 [00:00<00:00, 77709.75it/s]


Número de iteraciones con momentum (beta=0.4): 415


100%|██████████| 999/999 [00:00<00:00, 80041.83it/s]


Número de iteraciones con momentum (beta=0.5): 347


100%|██████████| 999/999 [00:00<00:00, 79796.41it/s]


Número de iteraciones con momentum (beta=0.6): 280


100%|██████████| 999/999 [00:00<00:00, 70787.25it/s]


Número de iteraciones con momentum (beta=0.7): 235


100%|██████████| 999/999 [00:00<00:00, 68601.48it/s]


Número de iteraciones con momentum (beta=0.8): 171


100%|██████████| 999/999 [00:00<00:00, 71284.62it/s]

Número de iteraciones con momentum (beta=0.9): 306


El numero de iteraciones del algoritmo con momentum es mayor al original en todos los valores de coeficientes con los que se relaizaron los calculos. (Al general un coeficiente de momentum >= 1 el algoritmo fallaba)

En comparacion al coeficiente original de momentum (0.6) con un numero total de 280 iteraciones no hay ningun coeficiente para el cual, el algoritmo mejorado converja el doble de rapido, dandose el numero minimo de iteraciones con un coeficiente de 0.8 para un total de 171 iteraciones.